### ISSUE TITLE
Quick Keys suggested fixes

### ISSUE THREAD

**Issue by flathat** [CORE-POS/IS4C#968]
* Version of CORE? 2.7, and master

* Issue with Office, Lane, or both? Lane, mostly.

* Is this [mostly] a bug report, feature request, or question? problem fix suggestions

Some problems or issues with suggested solutions:

#### 1. Buttons with images 
[`QuickKeys/quickkeys/quickkey.php L#83`](https://github.com/CORE-POS/IS4C/blob/master/pos/is4c-nf/plugins/QuickKeys/quickkeys/quickkey.php):    
In this implementation the image is the background with text over it. Without the text, from `value=`, the submit doesn't work.

```PHP
    function display($id=""){
        $ret = "";
        if ($this->img == ""){
       // ...
        } else {
            $ret .= sprintf("<input type=\"submit\"
                name=\"quickkey_submit\" id=\"%s\"
                value=\"%s\" class=\"quick_button\"
                style='background:url(\"%s\");' />",
                $id,
                $this->title,
                "quickkeys/imgs/".$this->img
            );
            // Instead of the current syntax, which doesn't display the image .
            $ret .= sprintf("<input type=\"submit\"
                name=\"quickkey_submit\" id=\"%s\"
                value=\"%s\" class=\"quick_button\"
                src=\"%s\" />",
                $id,
                $this->title,
               /* Isn't needed: MiscLib::base_url(). */
                "quickkeys/imgs/".$this->img
            );
            /* A separate statement, just for clarity in the above.
             *  No reason for it not to be combined in the same statement as
             * it currently is.
             */
            $ret .= sprintf("<input type=\"hidden\" name=\"%s\"
                value=\"%s\" />",
                md5($this->title),
                $this->output_text);
```
In this approach the image does not occupy the whole button. I didn't try much to solve that. But the submit works and there is no text over the image.
```PHP
            $ret .= sprintf('
                <button type="submit"
                    name="quickkey_submit" id="%s"
                    value="%s"
                    class="quick_button pos-button coloredBorder"><img src="%s" /></button>',
                $id,
                $this->title,
                "quickkeys/imgs/".$this->img
            );
```

#### 2. `QuickLookups` table doesn't have an image, `.img` field
To correspond with the third element of `new quickkey("Products","QK3","products.jpg")`

#### 3. Clear button in [`QKDisplay->preprocess()` L#99](https://github.com/CORE-POS/IS4C/blob/master/pos/is4c-nf/plugins/QuickKeys/QKDisplay.php#L99) doesn't work, i.e. exit back to `pos2.php`
I'm not certain this is the right way to handle the situation but it does let `Clear` work.
```PHP
       // In submit by Cancel `name=quickkey_submit` apparently doesn't exist.
        if (FormLib::get('quickkey_submit', false) !== false) {
            $output = "";
            $qstr = '';
            if (FormLib::get("clear") == 0) {
                // submit process changes line break
                // depending on platform
                // apostrophes pick up slashes
                $choice = str_replace("\r","",FormLib::get("quickkey_submit"));
                $choice = stripslashes($choice);

                $value = FormLib::get(md5($choice));

                $output = CoreLocal::get("qkInput").$value;
                $qstr = '?reginput=' . urlencode($output) . '&repeat=1';
                CoreLocal::set("currentid",CoreLocal::get("qkCurrentId"));
            }
            if (substr(strtoupper($output),0,2) == "QK"){
                CoreLocal::set("qkNumber",substr($output,2));

                return true;
            } else {
                // Submit by Cancel never gets here
                $this->change_page($this->page_url."gui-modules/pos2.php" . $qstr);
            }
            return false;
        }
        /* Add this
         */
        if (FormLib::get("clear") == 1) {
            $this->change_page($this->page_url."gui-modules/pos2.php" . '');
        }
```
#### 4. Argument for `QuickKeyLauncher()` in [`QKDisplay.php L#134](https://github.com/CORE-POS/IS4C/blob/master/pos/is4c-nf/plugins/QuickKeys/QKDisplay.php#L134)       
`$launcher = new QuickKeyLauncher($this->session);` or
`$launcher = new QuickKeyLauncher(array());`, since it doesn't seem to use any `$session` vars

#### 5. QK-aware error message if no keys defined
In [`QuickKeyLauncher->overlayKeys()` L#102 ](https://github.com/CORE-POS/IS4C/blob/master/pos/is4c-nf/plugins/QuickKeys/QuickKeyLauncher.php#L102) suggest a keys-oriented message.    
(This is what you see when `touchscreen` is first enabled and there are keys below the totals line on the POS but nothing has been assigned to them.)
```PHP
    private function overlayKeys($number, $preInput)
    {
        $my_keys = $this->getKeys($number);
        if (count($my_keys) == 0) {
            return DisplayLib::boxMsg(_('Key Set not found'), 
              '', false, DisplayLib::standardClearButton());
            // Instead of:
           return DisplayLib::boxMsg('Menu not found', 
               '', false, DisplayLib::standardClearButton());
```

**Comment by gohanman** [comment]
I think that covers the easier bits. I prefer "menu" for 5 since the same backend editor applies to both styles of menu. In either case though it'd be more helpful if the message told you **which** menu number it was looking for.

For images, I think it might make more sense to add a binary field to `QuickLookups` to store the actual images instead of just a file name (and probably another field for file type). Part of the advantage of putting menu entries into the table is they can be edited centrally and then pushed out to all relevant lanes. Storing just file names means there's an extra step to distribute new/updated files to the appropriate directory on each register. It's a minor hassle, but using the existing mechanisms for transferring database tables avoids it.

Meanwhile, 1 is addressable through CSS. Making the font size zero hides the text without removing it.

(Apparently putting "close" in a commit message closes an issue; didn't intend to do that)

**Comment by flathat** [comment]
> Making the font size zero hides the text without removing it.

I'm not sure having text visible is a bad thing; it may depend on how self-explanatory the image is.

**Comment by gohanman** [comment]
I agree in principle. But including any needed text in the image itself will give more control over exactly how it's placed/sized/etc.

**Comment by flathat** [comment]
A couple more bits on this:

#### 1. QK buttons: [quickkeys->display()L#83](https://github.com/CORE-POS/IS4C/blob/master/pos/is4c-nf/plugins/QuickKeys/quickkeys/quickkey.php#L83)    
+ Suppressing `value=` text, Firefox 58.0.1 ignores the `quickImageButton` class, which uses `font-size: 0;`, but does follow `style=` that does the same (`$more_style` below).  And added use of class `rounded`. 
Chrome 64.0 follows `quickImageButton` (but the result is that if the dimensions of the image are not perfect the "button" area is not filled).
+ The class `pos-button` is `button.pos-button`, only for `<button>`, and so can be removed.  If an identical class `input.pos-button` is defined it is ignored by Firefox but followed by Chrome.
```PHP
            $imgURL = 'quickkeys/' . (is_numeric($this->img)
                ? 'noauto/img.php?imgID=' . $this->img
                : 'imgs/' . $this->img);
            // Because quickButtonImage is ignored by Firefox 58
            $more_style = "font-size:0px;";
            $ret .= sprintf("<input type=\"submit\"
                name=\"quickkey_submit\" id=\"%s\" value=\"%s\"
                class=\"quick_button pos-button coloredBorder quickButtonImage rounded\"
                    style=\"background-image:url(%s); $more_style \" />
                <input type=\"hidden\" name=\"%s\"
                value=\"%s\" />",$id,$this->title,
                $imgURL,
                md5($this->title),
                $this->output_text);
```

#### 2. The [quickkeys->__constructor()L#55](https://github.com/CORE-POS/IS4C/blob/master/pos/is4c-nf/plugins/QuickKeys/quickkeys/quickkey.php#L55)
`Support for images on buttons instead of text is theoretical and not yet tested.`    
is no longer true, or at least not as true as it was.

#### 3. Add image support for overlay mode [`QuickLauncher->overlayKeys() L#110`](https://github.com/CORE-POS/IS4C/blob/master/pos/is4c-nf/plugins/QuickKeys/QuickKeyLauncher.php#L110)
+ Re `quickButtonImage`, `$more_style`, `pos-button`, `rounded` see comments above.
+ The code is presented as a draft but has been tested in this form.
```PHP
// ...
        /* Two new variables. Both could probably be literals.
         *  I was thinking of using $more_style to control display of the text on a case-by-case
         *   basis but it isn't implemented.
        */
        $pluginURL = '../plugins/QuickKeys/';
        $more_style = "font-size:0pt;";
        for ($i=0; $i<count($my_keys); $i++) {
            if ($i % 3 == 0) {
                if ($i != 0) {
                   $ret .= ' </div>';
                }
                $ret .= '<div class="qkRow">';
            }
            /* Add support for images on buttons.
             */
            if ($my_keys[$i]->img == ""){
                $ret .= sprintf('
                    <div class="qkBox">
                        <div id="qkDiv%d">
                            <button type="button" class="quick_button pos-button coloredBorder"
                                onclick="$(\'#reginput\').val($(\'#reginput\').val()+\'%s\');pos2.submitWrapper();">
                            %s A
                            </button>
                        </div>
                    </div>',
                    $i, $preInput . $my_keys[$i]->output_text, $my_keys[$i]->title);
            } else {
                $imgURL = $pluginURL . 'quickkeys/' . (is_numeric($my_keys[$i]->img)
                    ? 'noauto/img.php?imgID=' . $my_keys[$i]->img
                    : 'imgs/' . $my_keys[$i]->img);
                // (The three $ret .= could be in one sprintf()
                $ret .= sprintf('
                    <div class="qkBox">
                        <div id="qkDiv%d">', $i);
                $ret .= sprintf("<input value=\"%s\"
                    class=\"quick_button coloredBorder quickButtonImage rounded\"
                        onclick=\"$('#reginput').val($('#reginput').val()+'%s');pos2.submitWrapper();\"
                        style=\"background-image:url(%s); $more_style \" />",
                    $my_keys[$i]->title,
                    $preInput . $my_keys[$i]->output_text,
                    $imgURL);
                $ret .= '</div>
                    </div>';
            }
        }

```

**Comment by flathat** [comment]
A few comments or pointer to a primer on how this works would be welcome as I am not familiar with this level of HTML coding:
`quickkeys/noauto/img.php`
```PHP
<?php
use COREPOS\pos\lib\Database;
include(__DIR__ . '/../../../../lib/AutoLoader.php');
if (isset($_SERVER['HTTP_IF_MODIFIED_SINCE'])) {
    header('HTTP/1.1 304 Not Modified');
    return;
}
if (isset($_GET['imgID'])) {
    $dbc = Database::pDataConnect();
    $prep = $dbc->prepare('SELECT * FROM QuickLookups WHERE quickLookupID=?');
    $row = $dbc->getRow($prep, array($_GET['imgID']));
    if ($row) {
        header('Content-type: ' . $row['imageType']);
        header('Cache-control: max-age='.(60*60*24*365));
        header('Expires: '.gmdate(DATE_RFC1123,time()+60*60*24*365));
        header_remove('Pragma');
        echo $row['image'];
    }
}
```

**Comment by flathat** [comment]
Using the QuickLookups image editor I loaded an image BLOB in the record where `QuickLookups.quickLookupID = 1`, the same JPEG that worked from `quickkeys/imgs/foo.jpg` and have synced to the lane (using `$SYNC/special/generic.mysql.php`. On the lane, `.image` seems to be assigned, `[BLOB - 1.9  KiB]`, and `imageType` is `image/jpeg`.

In Chrome 64.0 (tried later, after writing the Firefox bit below) the image appears on the button as expected. Inspector shows `background-image: url(quickkeys/noauto/img.php?imgID=1);`

On the page in Firefox 58.0.1 the button is just a blue square, no image (but I can jigger to show the value="Member"). Inspecting in the browser shows:    
`<input name="quickkey_submit" id="qkButton0" value="Member" class="quick_button pos-button coloredBorder quickButtonImage rounded coloredArea" style="background-image:url(quickkeys/noauto/img.php?imgID=1);  " type="submit">`   
What might be going wrong? I have tried defeating the `HTTP_IF_MODIFIED_SINCE` test.

No error reported in `lane.log` or `debug_lane.log` or JS console.

This is irrelevant now:    
I notice that if I edit the record in Fannie, `QuickLookupsImages.php` the "Current Image" is broken, does not display. Inspecting shows `<img src="QuickLookupsImages.php?id=1">` and there is a message floating on it "Could not load the image".  Upon uploading again the image appears and in the Inspector the floating message displays the image.  I synced again, still no image on the lane. Opening the record in `QuickLookupsImages` now shows the image where before it didn't.

**Comment by gohanman** [comment]
That's strange. It works fine for me in both FF 58 and Chrome 64. Some section of the browser console should show network requests and HTTP response codes.

RE: `img.php`: it's just outputting an image. The `Content-type` is telling the browser it's an image. Everything else is telling the browser it's OK to cache the image. When I tried a menu with multiple images the delay in loading them all was noticeable.

**Comment by flathat** [comment]
> Some section of the browser console should show network requests and HTTP response codes.

Yes, in Firefox, that's the black-dot tab, "Net", which I clicked and enabled all of the options.  The request    
`https://foo.bar.coop/IS4C/pos/is4c-nf/plugins/QuickKeys/quickkeys/noauto/img.php?imgID=1`   has on the far right `[HTTP/1.1 200 OK 0ms]`, linked to the URL at the left.  When I take that in Firefox (or paste it into the address bar) there is a message: "cannot be displayed because it contains errors."  Pasting the URL into the Chrome address bar results in the image being displayed. IE 11 also displays it.

The image I was using (No.1) had been cropped, and possibly resized, using Microsoft Picture Manager.  Thinking that it might have been damaged in the manipulation I tried the original image (No.2).  That one displays in Firefox.  I then tried another image (No.3), that had been cropped from No.2 but I'm pretty sure not resized.  That one also displays.  So, Firefox may be more sensitive to file format irregularities than other browsers.  The image that is problematic for Firefox when stored-in/output-from the db displays fine when stored in the filesystem and accessed by regular URL.

**Comment by flathat** [comment]
Re the comment 5 above: "A couple more bits on this:"

No.1 Firefox is now applying the `class="quickButtonImage"` and also a new `input.pos-button` that is identical to `button.pos-button`.  Was partly a caching issue. Perhaps make `pos-button` not tag-specific?

No.3 I removed `$more_style` since it is no longer needed.  I *would* like to see the effect of this, that is images in overlays, in `master`.

**Comment by gohanman** [comment]
Yes, rendering issues were likely related to a cached version of `pos.css` that lacked `quickImageButton`.

I took a slightly different approach to images on overlaid keys, but pushing more of the display logic into `quickkey::display()` should keep the two display options for diverging as much.

**Comment by flathat** [comment]
It would be useful to have a field for the source of the binary in `QuickLookups.image`.  Display in the "Current Image" block.  

Also, I doubt it is useful for the Image field, the binary value in the Key/Menu Editor to be editable, maybe not even display it.

I seem to be having some trouble with the binaries becoming unreadable.  I just did Key/Menu Editor > "Save Changes" and now `img.php?ID=n` gets (Firefox) "cannot be displayed because it contains errors", for all three images that were on the page, Chrome and IE show placeholders.

**Comment by gohanman** [comment]
> It would be useful to have a field for the source of the binary in QuickLookups.image. Display in the "Current Image" block.

I'm not sure what source means here. The name of the original file?

> Also, I doubt it is useful for the Image field, the binary value in the Key/Menu Editor to be editable, maybe not even display it.

It isn't displayed - both because it's pointless noise for the user and because POST-ing the binary string and saving it back to the database as characters get escaped/encoded/etc through that process. Maybe that's the source of your errors? I didn't tag ff03dc5715cae53f92e4735267295f520d1bd04f with the issue number although it's semi-related. I try to keep plugin and non-plugin commits separate in case I want to cherry-pick them into an Office or plugin-specific repo.

**Comment by flathat** [comment]
> I'm not sure what source means here. The name of the original file?

Yes, "source" as in where it came from.

Re binary fields, sorry, I hadn't seen https://github.com/CORE-POS/IS4C/commit/ff03dc5715cae53f92e4735267295f520d1bd04f .  With that change the images are OK after editing the record.

**Comment by gohanman** [comment]
I don't understand what the point of the source field would be. The backend doesn't receive a filesystem path; it just gets a file name. If you know where that file is I don't see why you'd really need the name. If you don't know where the file is you can just download another copy from the browser instead of wasting time trying to find it.

**Comment by flathat** [comment]
If I had several similar versions of the image it's useful to know which one is actually being used.  It's less useful if the path isn't available.

**Comment by gohanman** [comment]
AFAIK no browsers send the client-side path with a file upload. I'm pretty
sure it's a security thing.

**Comment by gohanman** [comment]
This is all included in 2.9.0
